test/test_python_dispatch.py

```diff
@@ -562,6 +562,47 @@ def test_autograd_in_attr(self):
         self.assertIsNone(t.grad)
         self.assertIsNotNone(t.elem.grad)
 
+    def test_multiple_ops_subclass(self):
+        # This is a Direct Subclass, don't do that!
+        class MySubclass(torch.Tensor):
+            @staticmethod
+            def __new__(cls, elem):
+                r = torch.Tensor._make_subclass(cls, elem)
+                return r
+
+            __torch_function__ = torch._C._disabled_torch_function_impl
+
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+                with no_dispatch():
+                    return func(*args, **kwargs)
+
+        x = MySubclass(torch.rand(2, 2, dtype=torch.complex64))
+        y = x.conj()
+        # Details of the bug that this tests for:
+        # Here, y dispatch keys are: {PythonTLSSnapshot, AutogradCPU, Conjugate, Python, CPU}
+        # There are a few calls to the dispatcher that are going to happen here:
+        #  - call_exp: User calling exp on y
+        #    - PythonTLSSnapshot: records the TLS on entry and redispatch
+        #    - AutogradCPU: no input requires grad, so does nothing and redispatch
+        #    - Conjugate: no special implementation for exp: use the fallback that
+        #                 first clone the Tensor (to materialize the conj) then redispatch
+        #      - call_clone: conjugate fallback calling clone on y
```

Comment:
Potentially a stupid question, but, why does call_clone go to PythonTLSSnapshot? I would have expected the redispatch from Conjugate to go into the Python dispatch key?

Reply:
This is not a redispatch, this is a call. Namely: https://github.com/pytorch/pytorch/blob/f5e201e4e9fe946cd124ddcb4ed27b1b5b14386d/aten/src/ATen/native/MathBitsFallback.h#L102

Reply:
(I think that's because the conjugate fallback ends up calling `at::clone()`, which enters the dispatcher through `Dispatcher::call()`. PythonTLSSnapshot is set up to always run first every time you call `Dispatcher::call()`)

Reply:
Is the PythonTLSSnapshot key in the local exclude key set at this point? So I would expect the at::clone call to skip past PythonTLSSnapshot, AutogradCPU, and Conjugate

Reply:
Oh, I see, you don't use ExcludeDispatchKeyGuards to do the redispatch...

Reply:
My mental model of "when should you use `call()` vs `redispatch()` is basically: If you're a redispatching kernel, you should use `redispatch()` if you're redispatching for the *same* op, and `call()` if you're calling into a *different* op. That second case applies here, where the conjugate fallback for `exp()` needs to call into `clone()`, and it also applies for all composite ops.

Reply:
Ugh does this mean on the at::clone call, AutogradCPU doesn't get hit? AutogradCPU is in the local exclude set (because VariableType::blah both uses an exclude key guard AND at::redispatch)

Reply:
That is correct! My bad!

Reply:
Thanks for the detailed comment and explanations, I understand what's going on now. I agree that this test is testing the above code now.

On an orthogonal note, I'm a bit concerned that AutogradCPU doesn't get hit, but maybe that's not a problem because that's how things work even without the Python keys